### PR TITLE
feat: Add the Selection Sort Recursive algorithm

### PR DESCRIPTION
**Pull request by Tushar-K24**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1578"></a>

### CODE REVIEW

File: sorting/selection_sort_recursive.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{
+    test(); // execute the test
+    return 0;
+}
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:126]:
```suggestion
}

```

File: sorting/selection_sort_recursive.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    std::cout << "passed" << std::endl;
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{
+    test(); // execute the test
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:124]:
```suggestion
    test();  // run self-test implementations
```

File: sorting/selection_sort_recursive.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
+/**
+ * @file
+ * @brief Implementation of the [Selection
+ * sort](https://en.wikipedia.org/wiki/Selection_sort)
+ * implementation using recursion
+ * @details
+ * C++ program to sort an array using a recursive selection sort algorithm
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:7]: Please provide more details about the algorithm itself, how does it work, etc.. You can take the description of Wikipedia but adjust it to match our standards. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Woah, amazing work! 😄👍
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/Tushar-K24/C-Plus-Plus/actions

File: sorting/selection_sort_recursive.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+ * 
+ * Implementation
+ * FindMinIndex
+ * This function finds the minimum element of the array(list) recursively by simply comparing the minimum element of
+ * array reduced size by 1 and compares it to the last element of the array to find the minimum of the whole array. 
+ * 
+ * SelectionSortRecursive
+ * Just like selection sort, it divides the list into two parts (i.e. sorted and unsorted), and finds the minimum of the
+ * unsorted array by calling FindMinIndex, swaps the minimum element with the first element of the list and then
+ * solves recursively for the remaining unsorted list.
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:34]:
```suggestion
```

File: sorting/selection_sort_recursive.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
+/**
+ * @file
+ * @brief Implementation of the [Selection
+ * sort](https://en.wikipedia.org/wiki/Selection_sort)
+ * implementation using recursion
+ * @details
+ * C++ program to sort an array using a recursive selection sort algorithm
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:7]:
```suggestion
 * The selection sort algorithm divides the input list into two parts: a sorted sublist of items which is built up from 
 * left to right at the front (left) of the list, and a sublist of the remaining unsorted items that occupy the rest 
 * of the list. Initially, the sorted sublist is empty, and the unsorted sublist is the entire input list. The algorithm 
 * proceeds by finding the smallest (or largest, depending on the sorting order) element in the unsorted sublist, 
 * exchanging (swapping) it with the leftmost unsorted element (putting it in sorted order), and moving the sublist 
 * boundaries one element to the right.
 * 
 * ### Implementation
 * FindMinIndex
 * This function finds the minimum element of the array(list) recursively by simply comparing the minimum element of
 * array reduced size by 1 and compares it to the last element of the array to find the minimum of the whole array. 
 * 
 * SelectionSortRecursive
 * Just like selection sort, it divides the list into two parts (i.e.: sorted and unsorted) and finds the minimum of the
 * unsorted array. By calling the `FindMinIndex` function, it swaps the minimum element with the first element of the list, and then
 * solves recursively for the remaining unsorted list.
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: sorting/selection_sort_recursive.cpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+ * @brief Sorting algorithms
+ */
+namespace sorting {
+/**
+ * @namespace selection_sort_recursive
+ * @brief Functions for the [Selection
+ * sort](https://en.wikipedia.org/wiki/Selection_sort)
+ * implementation using recursion
+ */
+
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:42]:
```suggestion
```

File: sorting/selection_sort_recursive.cpp [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+namespace find_min_index {
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:43]:
```suggestion
namespace selection_sort_recursive {
```

File: sorting/selection_sort_recursive.cpp [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    if (current_position + 1 == in_arr.size()) {
+        return current_position;
+    }
+    uint64_t answer = findMinIndex(in_arr, current_position + 1);
+    if (in_arr[current_position] < in_arr[answer]) {
+        answer = current_position;
+    }
+    return answer;
+}
+}  // namespace find_min_index
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:62]:
```suggestion
```

File: sorting/selection_sort_recursive.cpp [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+namespace selection_sort_recursive {
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:64]:
```suggestion
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: sorting/selection_sort_recursive.cpp [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:63]:
```suggestion
```

File: sorting/selection_sort_recursive.cpp [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    if (current_position + 1 == in_arr.size()) {
+        return current_position;
+    }
+    uint64_t answer = findMinIndex(in_arr, current_position + 1);
+    if (in_arr[current_position] < in_arr[answer]) {
+        answer = current_position;
+    }
+    return answer;
+}
+}  // namespace find_min_index
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:61]:
```suggestion
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: sorting/selection_sort_recursive.cpp [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    // 1st test
+    // [1, 0, 2, 1] return [0, 1, 1, 2]
+    std::vector<uint64_t> array1 = {0, 1, 1, 2};
+    std::cout << "Test 1... ";
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:93]:
```suggestion
    std::cout << "1st test... ";
```

File: sorting/selection_sort_recursive.cpp [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    sorting::selection_sort_recursive::selectionSortRecursive(array1);
+    assert(std::is_sorted(std::begin(array1), std::end(array1)));
+    std::cout << "passed" << std::endl;
+    // 2nd test
+    // [1, 0, 0, 1, 1, 0, 2, 1] return [0, 0, 0, 1, 1, 1, 1, 2]
+    std::vector<uint64_t> array2 = {1, 0, 0, 1, 1, 0, 2, 1};
+    std::cout << "Test 2... ";
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:100]:
```suggestion
    std::cout << "2nd test... ";
```

File: sorting/selection_sort_recursive.cpp [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    sorting::selection_sort_recursive::selectionSortRecursive(array2);
+    assert(std::is_sorted(std::begin(array2), std::end(array2)));
+    std::cout << "passed" << std::endl;
+    // 3rd test
+    // [1, 1, 0, 0, 1, 2, 2, 0, 2, 1] return [0, 0, 0, 1, 1, 1, 1, 2, 2, 2]
+    std::vector<uint64_t> array3 = {1, 1, 0, 0, 1, 2, 2, 0, 2, 1};
+    std::cout << "Test 3... ";
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:107]:
```suggestion
    std::cout << "3rd test... ";
```

File: sorting/selection_sort_recursive.cpp [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    sorting::selection_sort_recursive::selectionSortRecursive(array3);
+    assert(std::is_sorted(std::begin(array3), std::end(array3)));
+    std::cout << "passed" << std::endl;
+    // 4th test
+    // [2, 2, 2, 0, 0, 1, 1] return [0, 0, 1, 1, 2, 2, 2]
+    std::vector<uint64_t> array4 = {2, 2, 2, 0, 0, 1, 1};
+    std::cout << "Test 4... ";
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:114]:
```suggestion
    std::cout << "4th test... ";
```

**Review by Panquesito7 — COMMENTED** [review]
Almost there! 😄

**Review by G3N3R4LE4RL — DISMISSED** [review]

File: sorting/selection_sort_recursive.cpp [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    // [1, 0, 0, 1, 1, 0, 2, 1] return [0, 0, 0, 1, 1, 1, 1, 2]
+    std::vector<uint64_t> array2 = {1, 0, 0, 1, 1, 0, 2, 1};
+    std::cout << "2nd test... ";
+    sorting::selection_sort_recursive::selectionSortRecursive(array2);
+    assert(std::is_sorted(std::begin(array2), std::end(array2)));
+    std::cout << "passed" << std::endl;
+    // 3rd test
+    // [1, 1, 0, 0, 1, 2, 2, 0, 2, 1] return [0, 0, 0, 1, 1, 1, 1, 2, 2, 2]
+    std::vector<uint64_t> array3 = {1, 1, 0, 0, 1, 2, 2, 0, 2, 1};
+    std::cout << "Test 3... ";
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:111]:
```suggestion
    std::cout << "3rd test... ";
```

File: sorting/selection_sort_recursive.cpp [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+    sorting::selection_sort_recursive::selectionSortRecursive(array3);
+    assert(std::is_sorted(std::begin(array3), std::end(array3)));
+    std::cout << "passed" << std::endl;
+    // 4th test
+    // [2, 2, 2, 0, 0, 1, 1] return [0, 0, 1, 1, 2, 2, 2]
+    std::vector<uint64_t> array4 = {2, 2, 2, 0, 0, 1, 1};
+    std::cout << "Test 4... ";
```
Comment (Panquesito7) [inline sorting/selection_sort_recursive.cpp:118]:
```suggestion
    std::cout << "4th test... ";
```

**Review by Panquesito7 — COMMENTED** [review]
Almost there! 😄

**Review by Panquesito7 — APPROVED** [review]
Awesome work, @Tushar-K24! It's been great for your first contribution here! Thank you for your contribution! 😄👍🎉

File: sorting/selection_sort_recursive.cpp [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+namespace selection_sort_recursive {
+/**
+ * @brief The main function finds the index of the minimum element
+ * @tparam T type of array
+ * @param in_arr array whose minimum element is to be returned
+ * @param current_position position/index from where the in_arr starts
+ * @returns index of the minimum element
+ */
+template <typename T>
+uint64_t findMinIndex(std::vector<T> &in_arr, uint64_t current_position = 0) {
```
Comment (ayaankhan98) [inline sorting/selection_sort_recursive.cpp:55]:
```suggestion
uint64_t findMinIndex(const std::vector<T> &in_arr, uint64_t current_position = 0) {
```

**Review by ayaankhan98 — APPROVED** [review]
LGTM :+1:

**Review by mishraabhinn — APPROVED** [review]

**Comment by Panquesito7** [comment]
@ayaankhan98, next time, please wait until all the requested reviewers review this PR. 🙂

**Comment by ayaankhan98** [comment]
> @ayaankhan98, next time, please wait until all the requested reviewers review this PR.

Ahh! no problem with that, I was just going with 2 reviewer policy.

**Comment by Panquesito7** [comment]
> > @ayaankhan98, next time, please wait until all the requested reviewers review this PR.
> 
> Ahh! no problem with that, I was just going with 2 reviewer policy. By the way we had 3 approved reviews on this PR.

Yeah. It is because a friend of mine will start helping with reviewing PRs now. Thanks anyways. 🙂

**Pull request merged**